runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,84 @@
+local G = require('_snippet_grammar')
+
+--- Returns the current cursor position (0-based).
+--- @return { line: integer, col: integer }
+local function cursor_position()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return { line = cursor[1] - 1, col = cursor[2] }
+end
+
+--- Resolves variables (like `$name` or `${name:default}`) as follows:
+--- - When a variable is unknown (i.e.: its name is not recognized in any of the cases below), return `nil`.
+--- - When a variable isn't set, return its default (if any) or an empty string.
+---
+--- Note that in some cases, the default is ignored since it's not clear how to distinguish an empty
+--- value from an unset value (e.g.: `TM_CURRENT_LINE`).
+---
+--- @param var string
+--- @param default? string
+--- @return integer|string|nil
+local function resolve_variable(var, default)
+  if var == 'TM_SELECTED_TEXT' then
+    return default or ''
+  elseif var == 'TM_CURRENT_LINE' then
+    return vim.api.nvim_get_current_line()
+  elseif var == 'TM_CURRENT_WORD' then
+    return vim.fn.expand('<cword>') --[[@as string]]
+      or default
+  elseif var == 'TM_LINE_INDEX' then
+    return cursor_position().line
```

Comment:
I'm unsure if `nvim_win_get_cursor` behaves like `getcursorpos` in it returning a zeroed position when the current position is invalid. In that case, I believe we should use the default here.